metric/gauge.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opencensus.io/internal/tagencoding"
+	"go.opencensus.io/metric/metricdata"
+)
+
+// Gauge represents a quantity that can go up an down, for example queue depth
+// or number of outstanding requests.
+//
+// Gauge maintains a value for each combination of of label values passed to
+// the Set or Add methods.
+//
+// Gauge should not be used directly, use Float64Gauge or Int64Gauge.
+type Gauge struct {
+	vals     sync.Map
+	desc     metricdata.Descriptor
+	start    time.Time
+	keys     []string
+	newEntry func() gaugeEntry
+}
+
+type gaugeEntry interface {
+	read(t time.Time) metricdata.Point
+}
+
+// Read returns the current values of the gauge as a metric for export.
+func (g *Gauge) Read() []*metricdata.Metric {
```

Comment:
Why is this public? Not sure I understand the design of these package. I was expecting to have a MetricsProducer (name is probably not the best) that implements the metricsexport.Producer, when create a new Gauge add it to a list of created Gauges in the MetricsProducer.

Reply:
I see, so basically you want it to work like it does in Java? I thought this would be a more natural Go API where you directly add the Gauge as a metricexport.Producer (this is why this method is public). I will make it work like Java instead by adding an intermediate MetricManager (or something like that).